### ISSUE TITLE
TFDF sets global state on async model execution

### ISSUE THREAD

**Issue by mattsoulanille** [tensorflow/tfjs#7571]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow.js installed from (npm or script link):
- TensorFlow.js version (use command below):
- Browser version:
- Tensorflow.js Converter Version:

**Describe the current behavior**
TFDF [sets global state when it executes an async model](https://github.com/tensorflow/tfjs/blob/master/tfjs-tfdf/src/tfdf_model.ts#L139). This [global state](https://github.com/tensorflow/tfjs/blob/8db53b0fafc565e45a0e1e4431732344f00a7160/tfjs-tfdf/src/register_tfdf_ops.ts#L25-L26) is used as an input to the `SimpleMLLoadModelFromPathWithHandle` (which also sets global state) and `SimpleMLInferenceOpWithHandle` ops.

This is a problem because async models can run at the same time as each other. It's possible that a model might be architected such that it sets the global state, yields to another task, and then reads the global state. This is a race condition because if the other task also sets the global state, the model will read a different global state and produce a bad result.

**Describe the expected behavior**
TFDF models should not rely on mutable global state.

One solution is to register a custom op for each TFDF model that refers to its own state instead of the global state. This still could have issues if `model.executeAsync` is called once and then again before the first run is finished, but it [looks like we're just storing the TFDF model in this state so we can run it during inference](https://github.com/tensorflow/tfjs/blob/8db53b0fafc565e45a0e1e4431732344f00a7160/tfjs-tfdf/src/register_tfdf_ops.ts#L61) (inference does not actually change the model). In this case, it would be fine.

This approach will let us remove the separate [`TFDFModel` class](https://github.com/tensorflow/tfjs/blob/master/tfjs-tfdf/src/tfdf_model.ts#L65) that wraps the TFJS `GraphModel` and just sets the global state before calling it because there will no longer be any global state to set (each model will store the state in its custom op).

Another approach is to store the assets and model runner in an constant node or a resource and add it as an input to the `SimpleMLLoadModelFromPathWithHandle` and `SimpleMLInferenceOpWithHandle` ops. This is probably better than the first approach I mentioned since it doesn't require registering custom ops.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/CodePen/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by pyu10055** [comment]
Looks like the asset and modelRunner should be in a lookup table

**Comment by avmohansai** [comment]
Hello @mattsoulanille   ,
Thank you for bringing this issue to our attention and for providing a comprehensive overview. We appreciate the time you invested in creating this issue.

Currently, our internal development team is concentrated on other scheduled milestones, so we cannot prioritize this task immediately. However, we would love to see this move forward as a community-led effort.

If you’d like to submit a fix, please fork the repo and open a PR following the [TensorFlow Contribution Guidelines](https://www.tensorflow.org/community/contribute), [Tfjs Contributions ](https://github.com/tensorflow/tfjs/blob/master/tfjs-vis/CONTRIBUTING.md#how-to-contribute)and be sure to link this issue in your PR to avoid duplicate work, kindly leave a comment here if you plan to take this on.

Thank you again for your interest and time.

**Comment by nileshau-afk** [comment]
Hello @mattsoulanille ,

Thank you for opening this issue. Since this issue has been open for a long time, we are closing this issue for now.
If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it.

Thank you for your understanding.
